File: subcommands/src/main/java/in/kyle/mcspring/subcommands/PluginCommandResolver.java
package in.kyle.mcspring.subcommands;

import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Optional;

import in.kyle.mcspring.command.CommandResolver;
import in.kyle.mcspring.command.Resolver;
import in.kyle.mcspring.command.SimpleMethodInjection;
import lombok.AllArgsConstructor;

@Component
@AllArgsConstructor
public class PluginCommandResolver implements CommandResolver {
    
    private final SimpleMethodInjection injection;
    
    @Override
    public Resolver makeResolver(Command command) {
        return parameter -> parameter.getType().isAssignableFrom(PluginCommand.class) ? Optional.of(
                makeCommand(command)) : Optional.empty();
    }
    
    private PluginCommand makeCommand(Command command) {
        ArrayList<String> args = new ArrayList<>(Arrays.asList(command.getArgs()));
        return new PluginCommand(injection, command.getSender(), args, new ArrayList<>());
    }
}

File: mcspring-api/mcspring-subcommands/src/main/java/in/kyle/mcspring/subcommands/PluginCommandResolver.kt
package `in`.kyle.mcspring.subcommands

import `in`.kyle.mcspring.command.CommandResolver
import `in`.kyle.mcspring.command.ParameterResolver
import `in`.kyle.mcspring.command.SimpleMethodInjection
import `in`.kyle.mcspring.subcommands.plugincommand.api.PluginCommand
import `in`.kyle.mcspring.subcommands.plugincommand.impl.PluginCommandImpl
import org.springframework.stereotype.Component

@Component
internal class PluginCommandResolver(
        private val injection: SimpleMethodInjection
) : CommandResolver {

    override fun makeResolver(command: CommandResolver.Command): ParameterResolver {
        return object : ParameterResolver {
            override fun resolve(parameter: Class<*>): Any? {
                return if (parameter.isAssignableFrom(PluginCommand::class.java)) {
                    return makeCommand(command)
                } else null
            }
        }
    }

    private fun makeCommand(command: CommandResolver.Command): PluginCommandImpl {
        return PluginCommandImpl(injection, command.sender, command.args.toMutableList(), true)
    }
}
